src/emitters/spot.cpp

```diff
@@ -0,0 +1,176 @@
+
+        m_cutoff_angle = deg_to_rad(m_cutoff_angle);
+        m_beam_width = deg_to_rad(m_beam_width);
+        m_inv_transition_width = 1.0f / (m_cutoff_angle - m_beam_width);
+        m_cos_cutoff_angle = cos(m_cutoff_angle);
+        m_cos_beam_width = cos(m_beam_width);
+        Assert(m_cutoff_angle >= m_beam_width);
+        m_uv_factor = tan(m_cutoff_angle);
+        m_texture = props.texture<Texture>("texture", Texture::D65(1.f));
+        m_flags = EmitterFlags::DeltaPosition | EmitterFlags::SpatiallyVarying;
+    }
+
+    inline Spectrum falloff_curve(const Vector3f &d, Float time, Wavelength wavelengths, Mask active) const {
+        SurfaceInteraction3f si = zero<SurfaceInteraction3f>();
+        si.wavelengths = wavelengths;
+        Spectrum result = m_intensity->eval(si, active);
+
+        const Transform4f &trafo = m_world_transform->eval(time, active);
+        Vector3f local_dir = trafo.inverse() * -d;
+        local_dir /= norm(local_dir);
+        const Float cos_theta = local_dir.z();
+
+        si.uv = Point2f(0.5f + 0.5f * local_dir.x() / (local_dir.z() * m_uv_factor), 0.5f + 0.5f * local_dir.y() / (local_dir.z() * m_uv_factor));
+        result *= m_texture->eval(si, active);
+        
+        return select(cos_theta <= m_cos_cutoff_angle, Spectrum(0.0f), select(cos_theta >= m_cos_beam_width, result, result * ((m_cutoff_angle - acos(cos_theta)) * m_inv_transition_width)));
+    }
+
+    std::pair<Ray3f, Spectrum> sample_ray(Float time, Float wavelength_sample,
+                                          const Point2f &spatial_sample,
+                                          const Point2f & /*dir_sample*/,
+                                          Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointSampleRay, active);
+
+        // 1. Sample spectrum
+        auto [wavelengths, spec_weight] = sample_wavelength<Float, Spectrum>(wavelength_sample);
+
+        // 2. Sample directional component
+        const Transform4f &trafo = m_world_transform->eval(time, active);
+        Vector3f d = trafo * warp::square_to_uniform_cone(spatial_sample, m_cos_cutoff_angle);
+
+        return { Ray3f(trafo * Point3f(0.f), d, time, wavelengths),
+                select(m_beam_width == m_cutoff_angle,
+                        unpolarized<Spectrum>(spec_weight) * 2 * math::Pi<Float> * (1 - m_cos_cutoff_angle),
+                        unpolarized<Spectrum>(spec_weight) * (2 * math::Pi<Float> * (1 - (sin(m_cutoff_angle) - sin(m_beam_width))/(m_cutoff_angle - m_beam_width)))) };
+    }
+
+    std::pair<DirectionSample3f, Spectrum> sample_direction(const Interaction3f &it,
+                                                            const Point2f & /*sample*/,
+                                                            Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointSampleDirection, active);
+
+        auto trafo = m_world_transform->eval(it.time, active);
+
```

Comment:
Here we should sample a direction on a cone and evaluate the corresponding pdf.

Please have a look at the Mitsuba 1 implementation, it should be very similar to that.

Reply:
I take it back, I got all confused with `sample_ray`. Sorry.

Reply:
Should the `ds.pdf` at line 133 be set to 1 like the implementations of point and directional emitters?

Reply:
Yes